Log handler functionality to persist and query local data

eliasyishak opened the pull request:
Functionality has been added to persist the events that have been sent to GA locally on the user's machine while also implementing a max number of logs to persist as defined by `kLogFileLength` in `lib/src/constants.dart`. A log handler method has also been added to query the persisted data for stats around usage on the user's machine, such as unique count of sessions, etc.

https://github.com/eliasyishak/dash_analytics/issues/4 – issue for persisting the data
https://github.com/eliasyishak/dash_analytics/issues/11 – issue for querying the locally persisted data

eliasyishak committed:
initializer creates the necessary log file to save events

eliasyishak committed:
first pass at algorithm to limit log file length

eliasyishak committed:
Merge pull request #19 from eliasyishak/main

Update with completed session handler branch

eliasyishak committed:
Update enums.dart

eliasyishak committed:
Merge pull request #21 from eliasyishak/main

Merging async updates in

eliasyishak committed:
update to test to include the log file

eliasyishak committed:
increasing log length + refactored logic to for adding data to log to catch edge case

By skipping the record we want to drop, we can ensure that the length of records remains capped at `kLogFileLength`

eliasyishak committed:
clean up

eliasyishak committed:
refactoring example file

eliasyishak committed:
Update dash_analytics_example.dart

eliasyishak committed:
querying functionality added to log handler

eliasyishak committed:
using enums for the tool name in the default analytics class

eliasyishak committed:
fix regex to look for underscores (no hyphens allowed)

eliasyishak committed:
Update dash_analytics_example.dart

eliasyishak committed:
updating enums to standardize toolOwner in DashEvents

eliasyishak committed:
fixing regex pattern to allow hyphens

eliasyishak committed:
test added to enforce rules on labels for DashTools

eliasyishak committed:
revert

eliasyishak committed:
fix typo

eliasyishak committed:
add'l documentation

eliasyishak committed:
test to added to check that logs are persisted and capped at limit

eliasyishak committed:
added test to check the sessionCount query on logs

eliasyishak committed:
cleanup to use data class instead of map for logfilestats

eliasyishak committed:
clean up

eliasyishak committed:
adding toString method for logFileStats

eliasyishak committed:
handle when the log file is empty

christopherfujino commented on lib/src/constants.dart:
```diff
@@ -55,6 +55,12 @@ reporting=1
 const String kDartToolDirectoryName = '.dart';
 
+/// How many data records to store in the log file
+const int kLogFileLength = 2500;
```
Do we really need this many?

eliasyishak replied:
I wanted to come in with a high value before we piloted it with the analyzer so I can see how many records are logged.

christopherfujino commented on lib/src/enums.dart:
```diff
@@ -4,31 +4,53 @@
+/// The tools that have been onboarded
+///
+/// All tools should use a hyphen as a delimiter
+enum DashTools {
```
The convention is to name enums singular

```suggestion
enum DashTool {
```

christopherfujino commented on lib/src/enums.dart:
```diff
@@ -4,31 +4,53 @@
   });
 }
 
+/// The tools that have been onboarded
```
```suggestion
/// Officially-supported clients of this package.
```

christopherfujino commented on lib/src/enums.dart:
```diff
@@ -4,31 +4,53 @@
 
+/// The tools that have been onboarded
+///
+/// All tools should use a hyphen as a delimiter
```
do you mean for the label?

eliasyishak replied:
Yes, good catch!

christopherfujino commented on lib/src/initializer.dart:
```diff
@@ -117,5 +123,12 @@ $tool=$dateStamp,$toolsMessageVersion
+    final File logFile = fs
+        .file(p.join(homeDirectory.path, kDartToolDirectoryName, kLogFileName));
+    if (!logFile.existsSync()) {
+      createLogFile(logFile: logFile);
```
do we have to create the file before actually writing to it (I'm just thinking better to do 1 file system write than 2, unless I'm missing something)

eliasyishak replied:
Yes I think in this case it would be good to initialize all of the necessary files the first time the analytics package is used. We are creating the other 3 files (session.json, CLIENT_ID, and config)

christopherfujino commented on lib/src/initializer.dart:
```diff
@@ -67,6 +67,12 @@ $tool=$dateStamp,$toolsMessageVersion
+  /// Creates that log file that will store the record formatted
+  /// events locally on the user's machine
+  void createLogFile({required File logFile}) {
+    logFile.createSync(recursive: true);
```
any reason not to make this async?

eliasyishak replied:
good point, there are several places where we do synchronous operations, I have created an issue for this so we can do them all at once instead of piece by piece

https://github.com/eliasyishak/dash_analytics/issues/26

christopherfujino commented on lib/src/log_handler.dart:
```diff
@@ -0,0 +1,99 @@
+  final int sessionCount;
+
+  /// Contains the data from the [LogHandler.logFileStats]
+  LogFileStats({
```
nit make a constructor const if it can be, then analyzer will be able to identify places where a construction can be marked const

```suggestion
  const LogFileStats({
```

christopherfujino commented on lib/src/log_handler.dart:
```diff
@@ -0,0 +1,99 @@
+  ///
+  /// It will be treated as an append only log and will be limited
+  /// to have has many data records as specified by [kLogFileLength]
+  LogHandler({
```
```suggestion
  const LogHandler({
```

christopherfujino replied:
Although, does this really need to be a class? Or can its methods just be methods of the `Analytics` class?

eliasyishak replied:
I prefer to have all the logic for the files this package interacts with in a class to make it easier to understand what is happening. I have done this with the config file with `ConfigHandler` and the session.json file also has `Session` to handle any session related operations

Also I am unable to make this a `const` constructor because the `logFile` class variable is not a constant value

christopherfujino replied:
> I prefer to have all the logic for the files this package interacts with in a class to make it easier to understand what is happening. I have done this with the config file with `ConfigHandler` and the session.json file also has `Session` to handle any session related operations

I think it's ultimately down to preference, so it's fine as is. I'm just pointing out the down-side to having a tree structure (where the `Analytics` class is the root node, and it has several dependent objects such as `ConfigHandler` and `LogHandler`) is that the end to end logic of (for example) sendEvent is less obvious, and you've created more boiler plate for yourself in terms of constructing all the objects.

> Also I am unable to make this a `const` constructor because the `logFile` class variable is not a constant value

Good point, I missed that.

eliasyishak replied:
That is a good point, we can look into possibly simplifying things in the future if we find that the log handler is only responsible for the save and logfilestats methods, but it could be a good idea now if we need add additional functionality as we onboard dash tools

christopherfujino commented on lib/src/analytics.dart:
```diff
@@ -281,13 +297,15 @@ class TestAnalytics extends AnalyticsImpl {
     // session file is getting updated without actually making any
     // POST requests to Google Analytics
-    generateRequestBody(
+    final Map<String, dynamic> body = generateRequestBody(
```
https://github.com/flutter/flutter/wiki/Style-guide-for-Flutter-repo#avoid-using-var-and-dynamic

```suggestion
    final Map<String, Object?> body = generateRequestBody(
```

christopherfujino replied:
And throughout this file, please :)

christopherfujino commented on lib/src/log_handler.dart:
```diff
@@ -0,0 +1,99 @@
+
+  /// Get stats from the persisted log file
+  LogFileStats? logFileStats() {
+    Iterable<Map<String, dynamic>> records =
```
```suggestion
    Iterable<Map<String, Object?>> records =
```

eliasyishak replied:
Got it. Just want to confirm if I refactored this correctly

OLD
```dart
    // Collection of unique sessions
    final Set<int> sessions = <int>{};
    for (Map<String, dynamic> element in records) {
      sessions.add(element['user_properties']['session_id']['value']);
    }
    final int sessionCount = sessions.length;
```

NEW
```dart
    // Collection of unique sessions
    final Set<int> sessions = <int>{};
    for (Map<String, Object?> element in records) {
      final Map<String, int> prop =
          element['user_properties'] as Map<String, int>;
      final int? value = prop['value'];
      if (value != null) sessions.add(value);
    }
    final int sessionCount = sessions.length;
```

christopherfujino replied:
This looks right. Can `element['user_properties']['session_id']['value']` ever be null?

christopherfujino replied:
Well, of course anything is possible. Should it ever be null?

eliasyishak replied:
As long as the log file hasn't been touched, there should always be a value for `session_id`. The structure of the json is determined by the schema needed by GA and a snippet is shown below

```
{
  'user_properties': {
    'session_id`: {
      'value': 12345678
    }
    ....
  }
}
```

So maybe instead of `int?`, I can make it `final int value = prop['value']!;`? and do away with the conditional check?

eliasyishak replied:
But now that I think I about it, is it safe to assume that any of the keys exist? Should I perform checks to make sure that `user_properties`, `session_id` and `value` exist? And if they don't just skip that record or return null from the `logFileStats()` method?

christopherfujino replied:
So with this kind of thing, there's always a balance of:

1. creating signal for ourselves that we have a bug
2. not disrupting what the user was doing

In this case, since this log persistence is not on the critical path of sending analytics events, I think it makes sense to skip the record if it is malformed (once we have this hooked up we'd obviously want to verify it's at least working locally for us).

christopherfujino commented on lib/src/log_handler.dart:
```diff
@@ -0,0 +1,99 @@
+  final Directory homeDirectory;
+  final File logFile;
+
+  /// This class is responsible for writing to a log
```
Dartdoc's for a class should be above the class definition (on line 34), and constructor dartdocs should describe that particular constructor's usage.

eliasyishak committed:
making enum for tools singular

eliasyishak committed:
dartdoc clean up for DashTool enum

eliasyishak committed:
making `LogFileStats` const constructor

eliasyishak committed:
update dartdocs for `LogHandler`

eliasyishak committed:
remove `dynamic` from analytics.dart

eliasyishak committed:
remove unused variable in example

eliasyishak committed:
making dashevents singular enum

eliasyishak committed:
Using a data class to parse each record item + add'l tests

This approach of using the `LogItem` class will ensure that when we are trying to perform analysis on the records, we will have all the data needed

The try/except block in the `LogItem.fromRecord` will return null if any of the records are malformed

eliasyishak commented on lib/src/log_handler.dart:
````diff
@@ -0,0 +1,225 @@
+  ///     }
+  /// }
+  /// ```
+  static LogItem? fromRecord(Map<String, dynamic> record) {
+    if (!record.containsKey('user_properties')) return null;
+
+    // Using a try/except here to parse out the fields if possible,
+    // if not, it will quietly return null and won't get processed
+    // downstream
+    try {
+      // Parse the data out of the `user_properties` value
+      final Map<String, dynamic> userProps = record['user_properties'];
+
+      // Parse out the values from the top level key = 'user_properties`
+      final int sessionId = userProps['session_id']['value'];
+      final String branch = userProps['branch']['value'];
+      final String host = userProps['host']['value'];
+      final String flutterVersion = userProps['flutter_version']['value'];
+      final String dartVersion = userProps['dart_version']['value'];
+      final String tool = userProps['tool']['value'];
+      final DateTime localTime =
+          DateTime.parse(userProps['local_time']['value']);
+
+      return LogItem(
+        sessionId: sessionId,
+        branch: branch,
+        host: host,
+        flutterVersion: flutterVersion,
+        dartVersion: dartVersion,
+        tool: tool,
+        localTime: localTime,
+      );
+    } catch (e) {
+      return null;
+    }
+  }
````
@christopherfujino let me know what you think about this approach – this will just return null for any malformed records

christopherfujino replied:
I think this will stop working if you change the parameter type from `dynamic` -> `Object?`, per the style guide. Also, it's not safe to catch everything, per this lint (which we have enabled in the tool) https://dart-lang.github.io/linter/lints/avoid_catches_without_on_clauses.html.

I know it's more verbose, but I think you should remove the dynamic and explicitly check for nulls.

eliasyishak replied:
That's fair, I've removed all instances of `dynamic` from the package code and cast them as the appropriate nullable type now

eliasyishak commented on lib/src/log_handler.dart:
```diff
@@ -0,0 +1,225 @@
+    // Parse each line of the log file through [LogItem],
+    // some returned records may be null if malformed, they will be
+    // removed later through `whereType<LogItem>`
+    final List<LogItem> records = logFile
+        .readAsLinesSync()
+        .map((String e) => LogItem.fromRecord(jsonDecode(e)))
+        .whereType<LogItem>()
+        .toList();
+
+    if (records.isEmpty) return null;
+
+    // Get the start and end dates for the log file
+    final DateTime startDateTime = records.first.localTime;
+    final DateTime endDateTime = records.last.localTime;
+
```
part 2 @christopherfujino  This is where it gets called, leaving us with a `List<LogItem>` that has no missing values

eliasyishak committed:
refactor to remove `dynamic` usage

eliasyishak committed:
removing all instances of dynamic for Object?

eliasyishak commented on lib/src/log_handler.dart:
```diff
@@ -0,0 +1,250 @@
+      final String? flutterVersion = (userProps['flutter_version']!
+          as Map<String, Object?>)['value'] as String?;
+      final String? dartVersion = (userProps['dart_version']!
+          as Map<String, Object?>)['value'] as String?;
+      final String? tool =
+          (userProps['tool']! as Map<String, Object?>)['value'] as String?;
+      final String? localTimeString = (userProps['local_time']!
+          as Map<String, Object?>)['value'] as String?;
+
+      // If any of the above values are null, return null since that
+      // indicates the record is malformed
+      final List<Object?> values = <Object?>[
+        sessionId,
+        branch,
+        host,
+        flutterVersion,
+        dartVersion,
+        tool,
+        localTimeString,
+      ];
+      for (Object? value in values) {
+        if (value == null) return null;
+      }
+
+      // Parse the local time from the string extracted
+      final DateTime localTime = DateTime.parse(localTimeString!);
+
+      return LogItem(
+        sessionId: sessionId!,
+        branch: branch!,
+        host: host!,
+        flutterVersion: flutterVersion!,
+        dartVersion: dartVersion!,
+        tool: tool!,
+        localTime: localTime,
+      );
+    } on Exception {
+      return null;
+    }
+  }
```
@christopherfujino I have refactored with specific casting for each value and also checked if any of them are null in a for loop, the structure of the expected json is also in the documentation for this method if you want to follow along

eliasyishak committed:
changing version to 0.0.1

christopherfujino commented on lib/src/log_handler.dart:
```diff
@@ -0,0 +1,250 @@
+        tool: tool!,
+        localTime: localTime,
+      );
+    } on Exception {
```
I don't think we need the `try` or this `on` clause, unless there is a specific `Exception` you're considering, in which case we should catch that specific Exception.

christopherfujino replied:
Otherwise, the rest LGTM

eliasyishak replied:
Would it be a `TypeError`? I was hoping to catch any cases of missing keys when grabbing values from the record `Map`

In the line below, I am saying that `session_id` has to exist.

```dart
      final int? sessionId =
          (userProps['session_id']! as Map<String, Object?>)['value'] as int?;
```

Or do you think it would be better to check the existence of all the keys (ie. `session_id`) before I grab the values.

christopherfujino replied:
I don't think a `TypeError` is an `Exception`, unfortunately :)

christopherfujino replied:
When I see this in crash, logging, it's a private Class `_CastError` we don't have access to. However, if `_CastError` is a `TypeError`, catching that should be doable.

eliasyishak replied:
Cool, I wanted to test this in dartpad and found that the `on` block was executed instead of the `catch` so I think switching it over may be alright

```dart
void main() {
  Map x = {};

  try {
    print(x['test']!);
  } on TypeError {
    print(123);  // this is getting printed
  } catch (e) {
    print(e);
  }
}

```

christopherfujino replied:
SGTM

eliasyishak committed:
catch errors for missing keys + malformed date strings
